File: Back Tracking/529.cpp
#include <bits/stdc++.h>

#define endl '\n'
#define vi vector<int>

using namespace std;

vi ans;

void back(int n, vi res, int &size);

int main(){
  int n;
  while(cin >> n, n){
    ans.clear();
    int size = 2147483647;
    back(n, vi(1, 1), size);
    cout << 1;
    for(int i = 1; i < size; ++i) cout << " " << ans[i];
    cout << endl;
  }
  return 0;
}

void back(int n, vi res, int &size){
  if(res.size() < size){
    if(res[res.size() - 1] == n){
      ans.clear();
      for(int i = 0; i < res.size(); ++i) ans.push_back(res[i]);
      size = res.size();
    }
    else if(res[res.size() - 1] < n){
      for(int i = res.size() - 1; i >= 0 && res[i] + res[i] > res[res.size() - 1]; --i){
        for(int j = i; j >= 0 && res[i] + res[j] > res[res.size() - 1]; --j){
          res.push_back(res[i] + res[j]);
          back(n, res, size);
          res.pop_back();
        }
      }
    }
  }
}

File: Back Tracking/529.py
from sys import stdin

size, ans = None, None

def back(n, res):
  global ans, size
  if len(res) < size:
    if res[-1] == n: ans, size = res.copy(), len(res)
    elif res[-1] < n:
      i = len(res) - 1
      while i >= 0 and res[i] + res[i] > res[-1]:
        j = i
        while j >= 0 and res[i] + res[j] > res[-1]:
          back(n, res + [res[i] + res[j]])
          j-=1
        i-=1

def main():
  global size, ans
  n, size, ans = int(stdin.readline().strip()), 2**32, []
  while n:
    back(n, [1])
    print(1, end="")
    for i in range(1, size): print(" " + str(ans[i]), end="")
    print()
    n, size, ans = int(stdin.readline().strip()), 2**32, []


main()